rust/tools/authorization-logic/src/ast.rs

```diff
@@ -54,58 +54,74 @@ impl PartialEq for AstPredicate {
 // See https://doc.rust-lang.org/std/cmp/trait.Eq.html
 impl Eq for AstPredicate {}
 
-#[derive(Debug, Clone, Serialize, Deserialize)]
+#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
```

Comment:
Discussed with @aferr about whether there was any risk (e.g. performance hit) to deriving these. I believe the answer is no, other than a small (likely negligible) increase to compile time and binary size. From [Rust docs](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html#derivable):

> When `derived` on structs, two instances are equal if all fields are equal, and not equal if any fields are not equal. When `derived` on enums, each variant is equal to itself and not equal to the other variants.

This implies that the implementation for `PartialEq` is a simple recursive equality check (for structs) and case match (for enums), which should be a fairly simple function. The [`Eq` trait](https://doc.rust-lang.org/std/cmp/trait.Eq.html#derivable) doesn't add any additional methods, it just tells Rust that the derived methods in `PartialEq` are actually full equivalence relations.

Informally, this [thread on Reddit](https://www.reddit.com/r/rust/comments/p5vsjl/is_there_any_reason_not_to_derive_every_possible/) seems to indicate the same, e.g. one user says:

> If you're exposing a struct that mostly represents a value-based bag of data, then you should be implementing/deriving Debug, Clone, PartialEq, Hash at a minimum. If you can derive Eq, you should do that, too.

Reply:
OK it sounds like it should be good to use this, then. Thanks for looking into it and for a thorough answer.